src/timely-util/src/reclock.rs

```diff
@@ -0,0 +1,958 @@
+//!         * _Proof: Predecessor exists because the set {t: t < t2} is non-empty since it must contain at least min(IntoTime)._
+//!     3. δ𝐓(t) = sum{δ𝐒(s): (𝐑(t2_prev) ⪯ s) ∧ !(𝐑(t) ⪯ s)}
+//!         * _Proof: By definition of δ𝐓(t)_
+//!     4. δ𝐒(s) ∈ δ𝐓(t1)
+//!         * _Proof: By <1>4._
+//!     5. Q.E.D
+//!         * _Proof: By <2>3 and <2>4._
+//! 9. 𝐑(t1) ⪯ 𝐑(t2_prev)
+//!     * _Proof: By <1>.7 and hypothesis on R_
+//! 10. 𝐑(t1) ⪯ s
+//!     * _Proof: By <1>8 and <1>9._
+//! 11. Q.E.D
+//!     * _Proof: By <1>5 and <1>10_
+
+use std::cmp::{Ordering, Reverse};
+use std::collections::binary_heap::{BinaryHeap, PeekMut};
+use std::collections::VecDeque;
+use std::iter::FromIterator;
+
+use differential_dataflow::difference::Semigroup;
+use differential_dataflow::lattice::Lattice;
+use differential_dataflow::{consolidation, AsCollection, Collection, ExchangeData};
+use mz_ore::collections::CollectionExt;
+use timely::communication::{Message, Pull, Push};
+use timely::dataflow::channels::pact::Pipeline;
+use timely::dataflow::operators::capture::Event;
+use timely::dataflow::operators::generic::builder_rc::OperatorBuilder;
+use timely::dataflow::operators::CapabilitySet;
+use timely::dataflow::Scope;
+use timely::order::{PartialOrder, TotalOrder};
+use timely::progress::frontier::{AntichainRef, MutableAntichain};
+use timely::progress::{Antichain, Timestamp};
+
+/// Constructs an operator that reclocks a `source` collection varying with some time `FromTime`
+/// into the corresponding `reclocked` collection varying over some time `IntoTime` using the
+/// provided `remap` collection.
+///
+/// In order for the operator to read the `source` collection a `Pusher` is returned which can be
+/// used with timely's capture facilities to connect a collection from a foreign scope to this
+/// operator.
+pub fn reclock<G, D, FromTime, IntoTime, R>(
+    remap_collection: &Collection<G, FromTime, i64>,
+    as_of: Antichain<G::Timestamp>,
+) -> (
+    Box<dyn Push<Message<Event<FromTime, (D, FromTime, R)>>>>,
+    Collection<G, D, R>,
+)
+where
+    G: Scope<Timestamp = IntoTime>,
+    D: ExchangeData,
+    FromTime: Timestamp,
+    IntoTime: Timestamp + Lattice + TotalOrder,
+    R: Semigroup,
+{
+    let mut scope = remap_collection.scope();
+    let mut builder = OperatorBuilder::new("Reclock".into(), scope.clone());
+    let info = builder.operator_info();
+    let channel_id = scope.new_identifier();
+    // Here we create a channel that can be used to send data from a foreign scope into this
+    // operator. The channel is associated with this operator's address so that it is activated
+    // every time events are availalbe for consumption. This mechanism is similar to Timely's input
+    // handles where data can be introduced into a timely scope from an exogenous source.
+    let (pusher, mut puller) =
+        scope.pipeline::<Event<FromTime, (D, FromTime, R)>>(channel_id, &info.address);
```

Comment:
It's a bit underspecified what this does when invoked by users for their goals. For example, I'd have to check to see if progress tracking derives from channels created and addressed somewhere, or if this prevents asserts around that structure.

In any case, perhaps it makes sense to drop both `info` and `channel` definitions under this comment as well, as afaict they are localized to this.